### PR TITLE
Gateway fetch contract mappings

### PR DESCRIPTION
**Pull request by suhasagg**
A new gateway get-contract-mapping <contract-address> command.
A new gateway list-contract-mappings command.

### CODE REVIEW

**Comment by suhasagg** [comment]
```
./loom gateway list-contract-mappings
{
  "confirmed_mapping": [
    {
      "from": "default:0x49F5C0FC8b951060604e145308EFffdC31D9bB3f",
      "to": "eth:0x9e51AEEeCa736CD81d27e025465834B8Ec08628A"
    },
    {
      "from": "default:0x85fb9206da279C05D007D5885dD0DFdD04257778",
      "to": "eth:0x1aA76056924bF4768D63357EcA6D6A56Ec929131"
    },
    {
      "from": "eth:0x1aA76056924bF4768D63357EcA6D6A56Ec929131",
      "to": "default:0x85fb9206da279C05D007D5885dD0DFdD04257778"
    },
    {
      "from": "eth:0x9e51AEEeCa736CD81d27e025465834B8Ec08628A",
      "to": "default:0x49F5C0FC8b951060604e145308EFffdC31D9bB3f"
    }
  ],
  "pending_mapping": []
}


./loom gateway get-contract-mapping 0x49F5C0FC8b951060604e145308EFffdC31D9bB3f
{
  "address": "eth0x9e51AEEeCa736CD81d27e025465834B8Ec08628A",
  "is_pending": false
}

```
Sample output for reference

File: cmd/loom/gateway/mapping_cmds.go [hunk 1]
```diff
@@ -265,6 +266,136 @@ func newMapAccountsCommand() *cobra.Command {
...
+			for _, mapping := range resp.PendingMappings {
+				pendingmappings = append(pendingmappings, getPendingMappingInfo(mapping))
+			}
+			mapping.PendingMappings = pendingmappings
+			mapping.ConfirmedMappings = confirmedmappings
+			output, err := json.MarshalIndent(mapping, "", "  ")
+			if err != nil {
+				return err
+			}
+			fmt.Println(string(output))
```
Comment (enlight) [inline cmd/loom/gateway/mapping_cmds.go]: Output the mapping in a table similar to how `chain-cfg list-features` outputs features https://github.com/loomnetwork/loomchain/blob/55eec93c0a19f08067197b5147e34d9d6b3461a3/cmd/loom/chainconfig/cmd.go#L229

So something like this:
```
from  | to    | status
--------------------------
0x... | 0x... | CONFIRMED
```
```

File: cmd/loom/gateway/mapping_cmds.go [hunk 2]
```diff
@@ -265,6 +266,136 @@ func newMapAccountsCommand() *cobra.Command {
...
+	cmd := &cobra.Command{
+		Use:     "get-contract-mapping <contract-addr>",
+		Short:   "Get Contract Mapping",
+		Example: getContractMappingCmdExample,
+		Args:    cobra.MinimumNArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var mapping Mapping
+			var contractAddr loom.Address
+			var err error
+			if strings.HasPrefix(args[0], "eth:") {
```
Comment (enlight) [inline cmd/loom/gateway/mapping_cmds.go]: This shouldn't be limited to `eth:`, we've also got `tron:` and `binance:`, and probably others, so lets make sure this works with any chain ID.

File: cmd/loom/gateway/mapping_cmds.go [hunk 3]
```diff
@@ -265,6 +266,136 @@ func newMapAccountsCommand() *cobra.Command {
...
+			req := &tgtypes.TransferGatewayGetContractMappingRequest{
+				From: contractAddr.MarshalPB(),
+			}
+			resp := &tgtypes.TransferGatewayGetContractMappingResponse{}
+			_, err = gateway.StaticCall("GetContractMapping", req, gatewayAddr, resp)
+			if err != nil {
+				return errors.Wrap(err, "failed to call gateway.GetContractMapping")
+			}
+			if resp.MappedAddress != nil {
+				mapping.Address = resp.MappedAddress.ChainId + ":" + resp.MappedAddress.Local.String()
```
Comment (enlight) [inline cmd/loom/gateway/mapping_cmds.go]: -> `loom.UnmarshalAddressPB(resp.MappedAddress).String()`

**Comment by suhasagg** [comment]
./loom gateway list-contract-mappings
```
From                                               | To                                                 | Status   
default:0x49F5C0FC8b951060604e145308EFffdC31D9bB3f | eth:0x9e51AEEeCa736CD81d27e025465834B8Ec08628A     | CONFIRMED
default:0x85fb9206da279C05D007D5885dD0DFdD04257778 | eth:0x1aA76056924bF4768D63357EcA6D6A56Ec929131     | CONFIRMED
eth:0x1aA76056924bF4768D63357EcA6D6A56Ec929131     | default:0x85fb9206da279C05D007D5885dD0DFdD04257778 | CONFIRMED
eth:0x9e51AEEeCa736CD81d27e025465834B8Ec08628A     | default:0x49F5C0FC8b951060604e145308EFffdC31D9bB3f | CONFIRMED
```

File: cmd/loom/gateway/mapping_cmds.go [hunk 4]
```diff
@@ -265,6 +266,117 @@ func newMapAccountsCommand() *cobra.Command {
...
+			gateway := client.NewContract(rpcClient, gatewayAddr.Local)
+			req := &tgtypes.TransferGatewayGetContractMappingRequest{
+				From: contractAddr.MarshalPB(),
+			}
+			resp := &tgtypes.TransferGatewayGetContractMappingResponse{}
+			_, err = gateway.StaticCall("GetContractMapping", req, gatewayAddr, resp)
+			if err != nil {
+				return errors.Wrap(err, "failed to call gateway.GetContractMapping")
+			}
+			if resp.MappedAddress != nil {
```
[the remark this answers is not in the document]
Comment (suhasagg) [reply cmd/loom/gateway/mapping_cmds.go]: Done.

File: cmd/loom/gateway/mapping_cmds.go (hunk 1 above)
Comment (suhasagg) [reply cmd/loom/gateway/mapping_cmds.go]: Done.

File: cmd/loom/gateway/mapping_cmds.go (hunk 2 above)
Comment (suhasagg) [reply cmd/loom/gateway/mapping_cmds.go]: Done.

File: cmd/loom/gateway/mapping_cmds.go (hunk 3 above)
Comment (suhasagg) [reply cmd/loom/gateway/mapping_cmds.go]: Done.

**Comment by enlight** [comment]
> 
> 
> ./loom gateway list-contract-mappings
> 
> ```
> From                                               | To                                                 | Status   
> default:0x49F5C0FC8b951060604e145308EFffdC31D9bB3f | eth:0x9e51AEEeCa736CD81d27e025465834B8Ec08628A     | CONFIRMED
> default:0x85fb9206da279C05D007D5885dD0DFdD04257778 | eth:0x1aA76056924bF4768D63357EcA6D6A56Ec929131     | CONFIRMED
> eth:0x1aA76056924bF4768D63357EcA6D6A56Ec929131     | default:0x85fb9206da279C05D007D5885dD0DFdD04257778 | CONFIRMED
> eth:0x9e51AEEeCa736CD81d27e025465834B8Ec08628A     | default:0x49F5C0FC8b951060604e145308EFffdC31D9bB3f | CONFIRMED
> ```

It's showing two entries for each mapping, should only show one.

**Comment by suhasagg** [comment]
Done. Have added functionality for unique mappings in the gateway contract ListMappings method

**Comment by suhasagg** [comment]
https://github.com/loomnetwork/transfer-gateway/pull/13

**Pull request merged**